File: Indexes of Subarray Sum.py
''' 
Given an unsorted array arr containing only non-negative integers, your task is to find a continuous subarray (a contiguous sequence of elements) whose sum equals a specified value target. 
You need to return the 1-based indices of the leftmost and rightmost elements of this subarray.
'''

def find_subarray(arr, target):
    current_sum = 0
    sum_map = {0: 0}  # To handle the case when subarray starts from index 0

    for i in range(len(arr)):
        current_sum += arr[i]

        if current_sum - target in sum_map:
            return sum_map[current_sum - target] + 1, i + 1  # 1-based indices

        sum_map[current_sum] = i + 1

    return -1, -1  # If no subarray is found

# Example usage
if __name__ == "__main__":
    arr = [1, 2, 3, 7, 5]
    target = 12
    result = find_subarray(arr, target)
    print("Subarray indices:", result)

File: Indexes of Subarray Sum.java
/* Given an unsorted array arr containing only non-negative integers, your task is to find a continuous subarray 
(a contiguous sequence of elements) whose sum equals a specified value target. 
You need to return the 1-based indices of the leftmost and rightmost elements of this subarray. */

class Solution {
    static ArrayList<Integer> subarraySum(int[] a, int k) {
        long s=0;
        for(int l=0,r=0;;){
            
            if(s<k){
                if (r<a.length){
                    // System.out.println(s+" +"+a[r]);
                    s+=a[r++];
                } else break;
            } else if (s>k){
                // System.out.println(s+" -"+a[l]);
                s-=a[l++];
            } else {
                return new ArrayList<Integer>(Arrays.asList(l+1,r));
            }
        }
        return new ArrayList<Integer>(Arrays.asList(-1));
    }
}
